vault-tool/src/Network/VaultTool/KeyValueV2.hs

```diff
@@ -0,0 +1,237 @@
+  where
+    path = vaultActionPath ReadSecretVersion mountedPath searchPath <> queryParams
+    queryParams = case version of
+        Nothing -> ""
+        Just n -> "?version=" <> T.pack (show n)
+
+newtype DataWrapper a = DataWrapper a
+
+instance ToJSON a => ToJSON (DataWrapper a) where
+    toJSON (DataWrapper x) = object ["data" .= x]
+
+instance FromJSON a => FromJSON (DataWrapper a) where
+    parseJSON = withObject "DataWrapper" $ fmap DataWrapper . (.: "data")
+
+{- | <https://www.vaultproject.io/docs/secrets/generic/index.html>
+
+ The value that you give must encode as a JSON object
+-}
+vaultWrite :: ToJSON a => VaultConnection Authenticated -> VaultSecretPath -> a -> IO ()
+vaultWrite conn (VaultSecretPath (mountedPath, searchPath)) = do
+    runVaultRequestAuthenticated_ conn
+        . withStatusCodes [200, 204]
+        . newPostRequest (vaultActionPath WriteSecret mountedPath searchPath)
+        . Just
+        . DataWrapper
+
+newtype VaultListResult = VaultListResult [Text]
+
+instance FromJSON VaultListResult where
+    parseJSON = withObject "VaultListResult" $ \v -> do
+        data_ <- v .: "data"
+        keys <- data_ .: "keys"
+        pure (VaultListResult keys)
+
+{- | <https://www.vaultproject.io/docs/secrets/generic/index.html>
+
+ This will normalise the results to be full secret paths.
+
+ Will return only secrets that in the are located in the folder hierarchy
+ directly below the given folder.
+
+ Use 'isFolder' to check if whether each result is a secret or a subfolder.
+
+ The order of the results is unspecified.
+
+ To recursively retrieve all of the secrets use 'vaultListRecursive'
+-}
+vaultList :: VaultConnection Authenticated -> VaultSecretPath -> IO [VaultSecretPath]
+vaultList conn (VaultSecretPath (VaultMountedPath mountedPath, VaultSearchPath searchPath)) = do
+    let path = vaultActionPath ListSecrets (VaultMountedPath mountedPath) (VaultSearchPath searchPath)
+    VaultListResult keys <-
+        runVaultRequestAuthenticated conn $
+            newListRequest path
+    pure $ map (VaultSecretPath . fullSecretPath) keys
+  where
+    fullSecretPath key = (VaultMountedPath mountedPath, VaultSearchPath (withTrailingSlash `T.append` key))
+    withTrailingSlash
+        | T.null searchPath = ""
+        | T.last searchPath == '/' = searchPath
+        | otherwise = searchPath `T.snoc` '/'
+
+{- | Recursively calls 'vaultList' to retrieve all of the secrets in a folder
+ (including all subfolders and sub-subfolders, etc...)
+
+ There will be no folders in the result.
+
+ The order of the results is unspecified.
+-}
+vaultListRecursive :: VaultConnection Authenticated -> VaultSecretPath -> IO [VaultSecretPath]
+vaultListRecursive conn location = do
+    paths <- vaultList conn location
+    flip concatMapM paths $ \path -> do
+        if isFolder path
+            then vaultListRecursive conn path
+            else pure [path]
+  where
+    concatMapM f xs = fmap concat (mapM f xs)
```

Comment:
You can eta reduce this to be `concatMapM f = fmap concat . mapM f`.  Also I would use `traverse` instead of `mapM` here.

Reply:
2c42064fa561f294710ccd8b3ca290ace3d42d3f